sdk/lib/drivers/wdf/wdfldr/class.c

```diff
@@ -0,0 +1,822 @@
+    _In_ PCLASS_MODULE ClassModule)
+{
+    if (ClassModule->ClassFileObject != NULL)
+    {
+        ObfDereferenceObject(ClassModule->ClassFileObject);
+        ClassModule->ClassFileObject = NULL;
+    }
+}
+
+
+VOID
+NTAPI
+ClassCleanupAndFree(
+    _In_ PCLASS_MODULE ClassModule)
+{
+    ClassClose(ClassModule);
+
+    if (ClassModule->Service.Buffer != NULL)
+    {
+        ExFreePoolWithTag(ClassModule->Service.Buffer, WDFLDR_TAG);
+        ClassModule->Service.Length = 0;
+        ClassModule->Service.Buffer = NULL;
+    }
+
+    if (ClassModule->ImageName.Buffer != NULL)
+    {
+        ExFreePoolWithTag(ClassModule->ImageName.Buffer, WDFLDR_TAG);
+        ClassModule->ImageName.Length = 0;
+        ClassModule->ImageName.Buffer = NULL;
+    }
+
+    ExDeleteResourceLite(&ClassModule->ClientsListLock);
+    ExFreePoolWithTag(ClassModule, WDFLDR_TAG);
+}
+
+
+PCLASS_MODULE
+NTAPI
+ClassCreate(
+    _In_ PWDF_CLASS_LIBRARY_INFO ClassLibInfo,
+    _In_ PLIBRARY_MODULE LibModule,
+    _In_ PUNICODE_STRING ServiceName)
+{
+    PCLASS_MODULE pNewClassModule;
+    NTSTATUS status;
+
+    pNewClassModule = ExAllocatePoolWithTag(NonPagedPool, sizeof(CLASS_MODULE), WDFLDR_TAG);
+
+    if (pNewClassModule == NULL)
+    {
+        goto exit;
+    }
+
+    RtlZeroMemory(pNewClassModule, sizeof(CLASS_MODULE));
+    if (ClassLibInfo != NULL)
+    {
+        pNewClassModule->ImplicitlyLoaded = TRUE;
+        pNewClassModule->ClassLibraryInfo = ClassLibInfo;
+        pNewClassModule->Version.Major = ClassLibInfo->Version.Major;
+        pNewClassModule->Version.Minor = ClassLibInfo->Version.Minor;
+        pNewClassModule->Version.Build = ClassLibInfo->Version.Build;
+    }
+
+    pNewClassModule->ClassRefCount = 1;
+    pNewClassModule->Library = LibModule;
+    ExInitializeResourceLite(&pNewClassModule->ClientsListLock);
+    InitializeListHead(&pNewClassModule->ClientsListHead);
+    InitializeListHead(&pNewClassModule->LibraryLinkage);
+    pNewClassModule->Service.Buffer = ExAllocatePoolWithTag(PagedPool, ServiceName->MaximumLength, WDFLDR_TAG);
+
+    if (pNewClassModule->Service.Buffer == NULL)
+    {
+        goto clean;
+    }
+
+    pNewClassModule->Service.MaximumLength = ServiceName->MaximumLength;
+    RtlZeroMemory(pNewClassModule->Service.Buffer, pNewClassModule->Service.MaximumLength);
+    RtlCopyUnicodeString(&pNewClassModule->Service, ServiceName);
+    status = GetImageName2(ServiceName, WDFLDR_TAG, &pNewClassModule->ImageName);
+
+    if (!NT_SUCCESS(status))
+    {
+        goto clean;
+    }
+
+    pNewClassModule->IsBootDriver = ServiceCheckBootStart(&pNewClassModule->Service);
+    status = GetImageBase(&pNewClassModule->ImageName, &pNewClassModule->ImageAddress, &pNewClassModule->ImageSize);
+
+    if (NT_SUCCESS(status))
+    {
+        return pNewClassModule;
+    }
+    else
+    {
+        __DBGPRINT(("GetImageBase failed\n"));
+    }
+
+clean:
+    ClassCleanupAndFree(pNewClassModule);
+exit:
+    return NULL;
+}
```

Comment:
This usage of goto block labels is ugly, especially when the `exit` block literally does nothing other than just returning NULL. Consider having one label block and use that for code path failures, cleanup, whatever.